### PR TITLE
feat: implement `stores-no-async`

### PR DESCRIPTION
**Pull request by baseballyama**
Related to: https://github.com/ota-meshi/eslint-plugin-svelte/issues/198 (`stores-no-async`)

This is the first ESLint rule I have written.
(So I'm sorry if I have a wrong understanding.)
I would be very grateful if you would review it.

### CODE REVIEW

File: docs/rules/no-store-async.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+---
+pageClass: "rule-details"
+sidebarDepth: 0
+title: "svelte/no-store-async"
+description: "disallow using async/await inside svelte stores"
+since: "v3.1.0"
```
Comment (baseballyama) [inline docs/rules/no-store-async.md:6]: We need to update here when we decide the release version of this rule.

File: src/rules/no-store-async.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      unexpected: "Do not pass async functions to svelte stores.",
+    },
+    type: "problem",
+  },
+  create(context) {
+    return {
+      CallExpression(node: ESTree.CallExpression) {
+        if (node.callee.type !== "Identifier") return
+        const { name } = node.callee
+        if (name !== "writable" && name !== "readable" && name !== "derived")
```
Comment (ota-meshi) [inline src/rules/no-store-async.ts:24]: Could you change it to use eslint-utils? By using eslint-utils, we can also check for cases like:

```js
import * as stores from "svelte/store"
stores.writable()
```

https://eslint-utils.mysticatea.dev/api/scope-utils.html#referencetracker-class

I think the implementation of require-stores-init is helpful.
https://github.com/ota-meshi/eslint-plugin-svelte/blob/3362ad4053c554ad54f7c1e42193282f7d2fdaac/src/rules/require-stores-init.ts#L21-L35

File: src/rules/no-store-async.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+          return
+        const [, fn] = node.arguments
+        if (fn.type !== "ArrowFunctionExpression" || !fn.async) return
```
Comment (ota-meshi) [inline src/rules/no-store-async.ts:27]: I think we should check if the argument exists.

```suggestion
        if (!fn || fn.type !== "ArrowFunctionExpression" || !fn.async) return
```

Comment (ota-meshi) [inline src/rules/no-store-async.ts:27]: I think we should also check `"FunctionExpression"` ( `async function () {}` ).
(Though I don't think people will use it much.)

https://github.com/estree/estree/blob/master/es5.md#functionexpression
https://github.com/estree/estree/blob/master/es2017.md#function

File: docs/rules/no-store-async.md (hunk 1 above)
Comment (ota-meshi) [reply docs/rules/no-store-async.md:6]: We don't need it now as it is automatically added by the release script.

```suggestion
```

File: .changeset/curvy-ants-admire.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+"eslint-plugin-svelte": minor
+---
+
+Add svelte/stores-no-async rule by @baseballyama in #225
```
Comment (ota-meshi) [inline .changeset/curvy-ants-admire.md:5]: It's probably automatically linked. However, we've just introduced changesets, so I'm not sure if it works yet 😅 

```suggestion
Add svelte/stores-no-async rule
```

**Review by ota-meshi — CHANGES_REQUESTED** [review]
Thank you for this PR!
I think it works mostly fine already, but I have some change requests to address some edge cases.

File: docs/rules/no-store-async.md (hunk 1 above)
Comment (baseballyama) [reply docs/rules/no-store-async.md]: removed!
https://github.com/ota-meshi/eslint-plugin-svelte/pull/225/commits/c05ef7cc01b234f3627bd4f2209214169884d474

File: .changeset/curvy-ants-admire.md (hunk 4 above)
Comment (baseballyama) [reply .changeset/curvy-ants-admire.md]: fixed!
https://github.com/ota-meshi/eslint-plugin-svelte/pull/225/commits/715d04a1b1cc2cf013ae8462afa4f1f2a392c92b

File: src/rules/no-store-async.ts (hunk 2 above)
Comment (baseballyama) [reply src/rules/no-store-async.ts]: updated!

https://github.com/ota-meshi/eslint-plugin-svelte/pull/225/commits/37d1de783bd7a6897a5f1b7377897576129ad485

I moved `extractStoreReferences` to utils folder but I'm not sure if this is your preferred way.

File: src/rules/no-store-async.ts (hunk 3 above)
Comment (baseballyama) [reply src/rules/no-store-async.ts]: fixed! Sorry for that!

https://github.com/ota-meshi/eslint-plugin-svelte/pull/225/commits/99e0996c1866e285e6d3d4d92172f037415e3500

Comment (baseballyama) [reply src/rules/no-store-async.ts]: fixed!

https://github.com/ota-meshi/eslint-plugin-svelte/pull/225/commits/0b8b4ce6273f987ba0746df5c852cdf0a9b6fd5b

**Comment by baseballyama** [comment]
@ota-meshi 

Thank you for your review!
I updated the PR according to your comments!
Could you please check it again? Thank you!

**Review by tivac — APPROVED** [review]

**Review by JounQin — APPROVED** [review]
I don't see anything strange. LGTM! 🍺

**Pull request merged**